sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PartitionScopedRegionLevelProgress.java

```diff
@@ -0,0 +1,318 @@
+                    return null;
+                }
+
+                this.normalizedRegionLookupMap.computeIfAbsent(regionRoutedTo, (regionRoutedToAsVal) -> regionRoutedToAsVal.toLowerCase(Locale.ROOT).trim().replace(" ", ""));
+
+                String normalizedRegionRoutedTo = this.normalizedRegionLookupMap.get(regionRoutedTo);
+
+                int regionId = RegionNameToRegionIdMap.getRegionId(normalizedRegionRoutedTo);
+
+                if (regionId != -1) {
+                    long localLsn = localLsnByRegion.v.getOrDefault(regionId, Long.MIN_VALUE);
+
+                    // regionId maps to a satellite region
+                    if (localLsn != Long.MIN_VALUE) {
+                        regionLevelProgressAsVal.compute(normalizedRegionRoutedTo, (normalizedRegionAsKey, regionLevelProgressAsValInner) -> {
+
+                            if (regionLevelProgressAsValInner == null) {
+                                return new RegionLevelProgress(parsedSessionToken.getLSN(), localLsn, null);
+                            }
+
+                            // regionLevelProgressAsValInner.parsedSessionToken is passed
+                            // to have a session token to merge with in case normalizedRegionRoutedTo
+                            // is equal to the first preferred region in the subsequent step
+                            request.requestContext.getSessionTokenEvaluationResults().add("Recording region specific progress of region : " + normalizedRegionRoutedTo + ".");
+                            return new RegionLevelProgress(
+                                Math.max(regionLevelProgressAsValInner.getMaxGlobalLsnSeen(), parsedSessionToken.getLSN()),
+                                Math.max(regionLevelProgressAsValInner.getMaxLocalLsnSeen(), localLsn),
+                                regionLevelProgressAsValInner.sessionToken);
+                        });
+                    }
+                    // regionId maps to a hub region
+                    else {
+                        regionLevelProgressAsVal.compute(normalizedRegionRoutedTo, (normalizedRegionAsKey, regionLevelProgressAsValInner) -> {
+
+                            if (regionLevelProgressAsValInner == null) {
+                                return new RegionLevelProgress(parsedSessionToken.getLSN(), Long.MIN_VALUE, null);
+                            }
+
+                            // regionLevelProgressAsValInner.parsedSessionToken is passed
+                            // to have a session token to merge with in case normalizedRegionRoutedTo
+                            // is equal to the first preferred region in the subsequent step
+                            request.requestContext.getSessionTokenEvaluationResults().add("Recording region specific progress of region : " + normalizedRegionRoutedTo + ".");
+                            return new RegionLevelProgress(
```

Comment:
I am delaying the merge until I have also identified the region is the first preferred readable region otherwise I just really need the `localLsn` and `globalLsn`. c: @xinlian12